### ISSUE TITLE
Upgrade to RN v0.64

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#4426]
The next round after #4245. This is the first time I'm opening one of these when we're already on the previous latest release, and before the new one is even out yet!

The RN community release managers are working on the next release now, and it's anticipated in the next few weeks. Details at https://github.com/react-native-community/releases/issues/214 and particularly at https://github.com/react-native-community/releases/issues/214#issuecomment-758203239 .

There's already a release candidate, so that would be a way to get an early check on any issues we'll run into.

**Comment by chrisbobbe** [comment]
We'll be jumping to Flow 0.137.0, which means (among other things) that something called ["types-first"](https://flow.org/en/docs/lang/types-first/) has been both introduced and enabled by default. We can turn it off, though, by adding `types_first=false` under `[options]` in our Flow config.

The thing we'll have to do to enable it is annotate almost all the exports of our modules:

> The caveat of this new version is that it requires exported parts of the code to be annotated with types, or to be expressions whose type can be trivially inferred (for example numbers and strings).

Not doing so for an export will raise a `signature-verification-failure` error.

It looks like we'll be able to work toward that incrementally by selecting certain directories whose exports should be checked.

They mention a [codemod](https://flow.org/en/docs/lang/types-first/#toc-a-codemod-for-large-codebases), but I suspect, if we use it at all, we'll want to use it on small sections of the code, one at a time, and check its work.

**Comment by gnprice** [comment]
> We'll be jumping to Flow 0.137.0, which means (among other things) that something called ["types-first"](https://flow.org/en/docs/lang/types-first/) has been both introduced and enabled by default

Cool! The performance benefits of that won't be as big for us as they are for a giant codebase like Facebook's, but we should see some. I think the quality of the type errors may improve, too.

Re the codemod, we'll definitely want to check its work: making readable commits and reviewing them normally. I suspect we'll want to start by trying it on relatively self-contained swathes of code like `src/api/` or even `src/third/`; maybe then reducers+selectors, the other files that don't know about React, and finally swathes of React components. Depending on how much manual work is needed (and how much effort is needed to review the output), we can scale up or down how much code we cover per commit.

And if the codemod is really bad at its job, we can always ignore it and do the changes manually. But I'm hopeful that it'll be good enough, automating enough boring stuff accurately enough, that it'll be worth using.

---

FTR here's the full list of changes we'll see -- up to this point in the changelog:
https://github.com/facebook/flow/blob/master/Changelog.md#01370
from 0.122.0, which we're currently on.

Other changes I see in a quick scan which we might appreciate, or might have to adapt for:

> * Allow indexers in exact object annotations

This should let us clean up a few types.

> * Standardized error suppression syntax, added ability to suppress errors based on error codes

The second part of this is #4433. The first means we'll have to drop `$FlowMigrationFudge` and replace it with one of the standard names.

> * Enable LSP support for autofix exports by default.
> * Enabled support for JSDoc in some LSP results.
> * Show JSDoc documentation in hover, autocomplete, and signature help.

I'm hopeful these will translate to some nice improvements in the VS Code editing experience!

> * Fixed Object.freeze to no longer incorrectly convert an inexact object into an exact object.

Hmm, I wonder if this will defeat the workaround we've used for a Flow bug involving `{}` as an "unsealed" object type.

**Comment by gnprice** [comment]
> We can turn it off, though, by adding `types_first=false` under `[options]` in our Flow config.

Oh also: this might be exactly the right thing to do for the RN upgrade. We'll want to do this if the types-first migration seems to be any substantial amount of work, I think.

But then we'll want to follow up by doing the types-first migration. A newer version of Flow (v0.143.0, out just this month) makes that the only available mode, so we'll presumably get that change in the *next* version of RN, v0.65.

**Comment by chrisbobbe** [comment]
And, v0.64 was released yesterday! Here's the blog post: https://reactnative.dev/blog/2021/03/11/version-0.64

**Comment by gnprice** [comment]
One small followup to do after this upgrade (or the relevant part of it): the upgraded Metro should fix the issue that #4819 worked around. So then we should revert the workaround.

**Comment by chrisbobbe** [comment]
Just opened #4843 to get to a version of react-native-webview whose peer-dep range for React Native include v0.64.

I didn't notice any other current third-party dependencies we'll have to do this for! 🎉

**Comment by chrisbobbe** [comment]
> > * Standardized error suppression syntax, added ability to suppress errors based on error codes
> 
> The second part of this is #4433.

Hmm, looks like RN v0.64 was released without adding all the needed suppressions, so we'll get dozens of warnings like this:

```
Warning ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ node_modules/react-native/index.js:651:3

Suppression is missing a code. Please update this suppression to use an error code: $FlowFixMe[prop-missing]

     648│     },
     649│   });
     650│
     651│   // $FlowFixMe This is intentional: Flow will error when attempting to access CheckBox.
     652│   Object.defineProperty(module.exports, 'CheckBox', {
     653│     configurable: true,
     654│     get() {
```

Looks like there was a commit to address this, facebook/react-native@bac2c2c80, released in RN v0.65.

**Comment by chrisbobbe** [comment]
I've just opened https://github.com/zulip/zulip-mobile/issues/4890 in our repo and send https://github.com/expo/expo/pull/13549 to `jest-expo` to satisfy some peer-dep constraints.

**Comment by chrisbobbe** [comment]
> > * Fixed Object.freeze to no longer incorrectly convert an inexact object into an exact object.
> 
> Hmm, I wonder if this will defeat the workaround we've used for a Flow bug involving `{}` as an "unsealed" object type.

It doesn't seem to, experimentally. I think https://github.com/facebook/flow/commit/059fbaf92e7cb76656aaff077ee63c08333a6c8a is the commit where this supposedly happened. I'm not really following what's going on in the OCaml code there.

**Comment by gnprice** [comment]
Since we're about to merge #4991 to complete this upgrade :tada: , I just scanned this thread to look for follow-ups. It looks like there's just one:

> One small followup to do after this upgrade (or the relevant part of it): the upgraded Metro should fix the issue that #4819 worked around. So then we should revert the workaround.
